### PR TITLE
Grab bag of Dates fixes

### PR DESCRIPTION
**Pull request by quinnj**

### CODE REVIEW

File: stdlib/Dates/src/types.jl [hunk 1]
```diff
@@ -350,6 +350,7 @@ function ==(a::Time, b::Time)
         microsecond(a) == microsecond(b) && nanosecond(a) == nanosecond(b)
 end
 (==)(x::TimeType, y::TimeType) = (===)(promote(x, y)...)
+Base.hash(x::Time, h::UInt) = hash(hour(x), hash(second(x), hash(second(x), h)))
```
Comment (JeffBezanson) [inline stdlib/Dates/src/types.jl]: Did you mean to include `minute` instead of `second` twice?

**Comment by JeffBezanson** [comment]
These commits will have to be separated. The hash change is a bug fix (whether it's suitable for a point release is debatable, but it can definitely go in 1.1), but the deprecation will take longer. For now I'd recommend moving the deprecated methods to deprecated.jl as you've done, but with no warnings for now, until we decide how to handle them.

File: stdlib/Dates/src/types.jl (hunk 1 above)
Comment (quinnj) [reply stdlib/Dates/src/types.jl]: Yeah, I was having source build problems, so I didn't get a chance to try any of this out locally. I think I've got it sorted out now.

**Comment by quinnj** [comment]
Ok, I commented out the `depwarn`s for now.

File: stdlib/Dates/src/deprecated.jl [hunk 2]
```diff
@@ -1,3 +1,67 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+using Base: @deprecate, depwarn
+
 # 1.0 deprecations
+function (+)(x::AbstractArray{<:TimeType}, y::GeneralPeriod)
+    # depwarn("non-broadcasted arithmetic is deprecated for Dates.TimeType; use broadcasting instead", nothing)
```
Comment (fredrikekre) [inline stdlib/Dates/src/deprecated.jl]: All the `nothing`s should be updated too?

Comment (vtjnash) [reply stdlib/Dates/src/deprecated.jl]: the name of the function that's being deprecated (in this case `:(+)`)

also mark all functions that contain a call to `depwarn` with `@noinline` so we can print the warning with correct location info

**Comment by JeffBezanson** [comment]
This seems fine for now. The details of the deprecations can be settled when we decide to active them.

**Comment by KristofferC** [comment]
Is the first commit backportable?

**Comment by JeffBezanson** [comment]
I think so; this fixes the hash function to obey the correctness property, so it's definitely a bug fix.

**Comment by KristofferC** [comment]
Ok, it seems the first commit is already included in 1.0.2: https://github.com/JuliaLang/julia/pull/29742.

**Pull request merged**